File: python/examples/parent_connects.py
#!/usr/bin/env python3
"""
FILE: python/examples/parent_connects.py
PURPOSE: Async v5 caller example that connects to the named math service.
"""

from __future__ import annotations

import asyncio
from pathlib import Path

from multifrost import connect, spawn


async def main() -> None:
    service = spawn(Path(__file__).with_name("math_worker_service.py"))
    connection = connect("math-service")
    handle = connection.handle()

    await handle.start()
    try:
        print(await handle.call.add(5, 3))
        print(await handle.call.multiply(4, 7))
        print(await handle.call.factorial(5))
    finally:
        await handle.stop()
        service.stop()
        service.wait()


if __name__ == "__main__":
    asyncio.run(main())

File: javascript/examples/parent_connects.ts
/**
 * FILE: javascript/examples/parent_connects.ts
 * PURPOSE: Show multiple Node callers connecting to the same v5 service peer through the router.
 * OWNS: Caller-side connect/handle usage for concurrent remote calls.
 * EXPORTS: None
 * DOCS: docs/spec.md, javascript/README.md
 */

import { connect } from "../src/index.js";

async function callerTask(callerId: number): Promise<void> {
    const connection = connect("math-service", {
        eagerTargetQuery: "exists",
    });

    const handle = connection.handle();
    await handle.start();

    try {
        for (let iteration = 0; iteration < 3; iteration += 1) {
            const left = callerId * 10;
            const result = await handle.call.add(left, iteration);
            console.log(`caller ${callerId}: ${left} + ${iteration} = ${result}`);
        }
    } finally {
        await handle.stop();
    }
}

async function main(): Promise<void> {
    await Promise.all([callerTask(1), callerTask(2), callerTask(3)]);
}

void main().catch(error => {
    console.error(error);
    process.exitCode = 1;
});
